### ISSUE TITLE
Ability to do emergency releases with comprehensive CI

### ISSUE THREAD

**Issue by daira** [zcash/zcash#2497]
We'd like to be able to quickly produce an emergency security/stability release that does not necessarily include everything added to master since the previous release.

Currently this runs into the obstacle that zkbot can only do tries and merges on master.

**Comment by daira** [comment]
We can currently trigger buildbot manually on a branch, so that may work as a temporary workaround.

**Comment by nathan-at-least** [comment]
For this ticket, let's exclude homu support and do that in a followup ticket.

Given that, let's document the "manual" process and then do a dry-run on the dev-ci system.

**Comment by str4d** [comment]
An emergency release of this kind would necessarily happen on its own branch on the main repository. There may be a difference in whether that branch subsequently is merged back into master or not (if the fix is just applied on top of the last release, then in nearly all cases it would be merged back in afterwards; if we need to cherry-pick things from master then we may not want to merge the branch back in), but that is irrelevant for the release process.

Here's a rough outline of what I think the process would be:

- A developer creates a hotfix branch from the previous release tag. They push this branch to the main repository.
- Developers now create the hotfix PRs based on (and merging into) this branch.
  - Other developers review and ACK.
  - A CI merge build is manually run, by force-triggering the `pr-merge` builder with the following fields set:
    - Repository: `https://github.com/<DevUser>/zcash`
      - `<DevUser>` must be in the set of "safe" users as-specified in the CI config.
    - Branch: name of the hotfix branch.
  - A link to the build and its result is manually added to the PR as a comment.
  - If the build was successful, the PR is merged via the GitHub button.
- Once the hotfix branch is complete, a release is conducted as-normal.
  -  Using the hotfix branch instead of the master branch may require modifications to the release script.
  - The release PR is merged following the above steps.
- Post-release, a decision can be made about merging the hotfix branch into master. If done, it would be via the usual PR merge process and Homu.

I've included PRs in the process (instead of just working in one branch) because "hotfix Homu" would require them, and it more closely matches our existing process. However, by doing so, note that the CI builds are on the PR branch being merged, *not* on the post-merge code - so the CI is not testing the interaction of the PR with later code in the branch. This is not a huge loss, because for a hotfix there will likely only be a few PRs, and they can be implemented sequentially. The final release process CI build will also catch any errors resulting from PRs being combined, as it is by definition conducted from the result of merging all PRs.

**Comment by daira** [comment]
If the hotfix branch is in the main repo, wouldn't the URL for the merge build just be https://github.com/zcash/zcash ? 

> Using the hotfix branch instead of the master branch may require modifications to the release script.

I consider that within the scope of this ticket (i.e. adding a branch argument to the script that defaults to ``master``).

**Comment by daira** [comment]
Also, doing a dry run is in the scope of this ticket.

**Comment by daira** [comment]
ACK on the process above, but a dry run is not going to get done before the 1.0.11 release. Let's do it soon after the release though.

**Comment by str4d** [comment]
> If the hotfix branch is in the main repo, wouldn't the URL for the merge build just be https://github.com/zcash/zcash ?

In my process above, we'd be merging developer branches *into* the hotfix branch, so we want to test the developer branches.

**Comment by str4d** [comment]
I'm planning another requirement: the hotfix branch must be named `hotfix-[RELEASE]` e.g. (`hotfix-v1.0.11-1`). That enables us to enforce the correct branch in the release script.

**Comment by str4d** [comment]
I have implemented the release process, and tested it on dev-ci.
